### PR TITLE
destination-s3 use instanceprofile if credentials are not provided

### PR DESCRIPTION
**Pull request by alvaroqueiroz**
## What
*Describe what the change is solving*
Destination-s3 connector does not support using instanceprofile authentication. This is a huge drawback for some organizations.

This solves the issues [5942](https://github.com/airbytehq/airbyte/issues/5942) and [8227](https://github.com/airbytehq/airbyte/issues/8227)

## How
Allow the fields accessKeyId and secretAccessKey to be optional on the destination-s3 configuration page.
If they are not provided, the getS3Client method will use InstanceProfileCredentialsProvider.

effect:
image

build:
cmd used - `./gradlew :airbyte-integrations:connectors:destination-s3:build`
image

result of sync, Airbyte is running on ec2 instance, and no aws credentials were provided
image

my modified connector image can be found [here](https://hub.docker.com/r/alvaroqueirozrs/airbyte-destination-s3)

## Recommended reading order
doc for [InstanceProfileCredentialsProvider](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/InstanceProfileCredentialsProvider.html)

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

Users will now be able to use instance profile authentication for the connector destination-s3

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>

#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [x] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
thanks @alvaroqueiroz

File: airbyte-integrations/connectors/destination-s3/src/main/resources/spec.json [hunk 1]
```diff
@@ -12,8 +12,6 @@
       "s3_bucket_name",
       "s3_bucket_path",
       "s3_bucket_region",
-      "access_key_id",
-      "secret_access_key",
```
[the remark this answers is not in the document]
Comment (alvaroqueiroz) [reply airbyte-integrations/connectors/destination-s3/src/main/resources/spec.json:16]: done

File: docs/integrations/destinations/s3.md [hunk 2]
```diff
@@ -223,6 +223,7 @@ Under the hood, an Airbyte data stream in Json schema is first converted to an A
 
 | Version | Date | Pull Request | Subject |
 | :--- | :--- | :--- | :--- |
+| 0.2.3 | 2022-01-10 | [\#9399](https://github.com/airbytehq/airbyte/pull/9399) | Use instance profile authentication if credentials are not provided |
```
[the remark this answers is not in the document]
Comment (alvaroqueiroz) [reply docs/integrations/destinations/s3.md:229]: done!

File: airbyte-integrations/connectors/destination-s3/src/main/resources/spec.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (alvaroqueiroz) [reply airbyte-integrations/connectors/destination-s3/src/main/resources/spec.json:16]: I think that the user-experience will not suffer,

if access_key_id AND secret_access_key are not provided -> instanceprofile auth
if access_key_id OR secret_access_key are not provided -> standard authentication error
the rest will stay the same.

**Comment by sherifnada** [comment]
reassigning to @edgao for review

File: airbyte-integrations/connectors/destination-s3/src/main/resources/spec.json [hunk 3]
```diff
@@ -72,7 +70,7 @@
       },
       "access_key_id": {
         "type": "string",
-        "description": "The access key id to access the S3 bucket. Airbyte requires Read and Write permissions to the given bucket.",
+        "description": "The access key id to access the S3 bucket. Airbyte requires Read and Write permissions to the given bucket, if not set, Airbyte will rely on Instance Profile",
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-s3/src/main/resources/spec.json:73]: this should also mention that if acces_key_id is set then secret_key_id must also be provided

File: docs/integrations/destinations/s3.md [hunk 4]
```diff
@@ -199,7 +199,7 @@ Under the hood, an Airbyte data stream in Json schema is first converted to an A
 #### Requirements
 
 1. Allow connections from Airbyte server to your AWS S3/ Minio S3 cluster \(if they exist in separate VPCs\).
-2. An S3 bucket with credentials.
+2. An S3 bucket with credentials or a instanceprofile with read/write permissions configured for the host(ec2, eks).
```
Comment (edgao) [inline docs/integrations/destinations/s3.md:202]:
```suggestion
2. An S3 bucket with credentials or an instanceprofile with read/write permissions configured for the host (ec2, eks).
```

File: airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/S3DestinationConfig.java [hunk 5]
```diff
@@ -128,7 +129,13 @@ public S3FormatConfig getFormatConfig() {
   public AmazonS3 getS3Client() {
     final AWSCredentials awsCreds = new BasicAWSCredentials(accessKeyId, secretAccessKey);
 
-    if (endpoint == null || endpoint.isEmpty()) {
+    if (accessKeyId == null && secretAccessKey == null || accessKeyId.isEmpty() && secretAccessKey.isEmpty()) {
+      return AmazonS3ClientBuilder.standard()
+      .withCredentials(new InstanceProfileCredentialsProvider(false))
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/S3DestinationConfig.java:134]: nitpick: you should format this before merging (maybe just run the `gradlew format` command)

File: airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/S3DestinationConfig.java [hunk 6]
```diff
@@ -87,8 +88,8 @@ public static S3DestinationConfig getS3DestinationConfig(final JsonNode config)
         config.get("s3_bucket_name").asText(),
         bucketPath,
         config.get("s3_bucket_region").asText(),
-        config.get("access_key_id").asText(),
-        config.get("secret_access_key").asText(),
+        config.get("access_key_id") == null ? "" : config.get("access_key_id").asText(),
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/S3DestinationConfig.java:91]: let's explicitly check that these are both set or both null, and throw an error if exactly one is non-null.

**Review by edgao — CHANGES_REQUESTED** [review]
added a few minor comments!

File: airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/S3DestinationConfig.java (hunk 6 above)
Comment (alvaroqueiroz) [reply airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/S3DestinationConfig.java:91]: Done, if the user inputs only  id OR keys, a message with error will be presented

image

File: airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/S3DestinationConfig.java (hunk 5 above)
Comment (alvaroqueiroz) [reply airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/S3DestinationConfig.java:139]: sorry about that, this line did not seem that long here

File: airbyte-integrations/connectors/destination-s3/src/main/resources/spec.json (hunk 3 above)
Comment (alvaroqueiroz) [reply airbyte-integrations/connectors/destination-s3/src/main/resources/spec.json]: done

**Review by edgao — APPROVED** [review]
one small wording change, otherwise LGTM!

**Review by alvaroqueiroz — COMMENTED** [review]
exception was added for the case when only id OR secret is inputed

**Comment by alvaroqueiroz** [comment]
@marcosmarxm and @sherifnada is anything else needed to proceed with the merge?

**Review by marcosmarxm — APPROVED** [review]
thanks @alvaroqueiroz

**Pull request merged**